README.md

```diff
@@ -306,6 +307,19 @@ Some servers may support certain helpers and not others. The AEP and associated
 
 Similarly, matching behavior will be determined by each implementing system. For example, some systems may match people based on email address or other information.
 
+### Deviations from RESTful Behavior
+
+This section outlines areas where the expectations of the OSDI customer community differ from boilerplate RESTful behavior and specify best practice expected behavior.
+
+#### Merging Objects (hashes) and Arrays on update
+
+Numerous resources in OSDI contain embedded objects (hashes) and arrays. For example, on Person, Birthday and Custom Fields are Objects (hashes) and Postal Addresses is an array.
+
+When dealing with updates to these elements of resources, servers should merge rather than replace these elements.  This is more consistent with expected user behavior.  For example, sending a person_signup_helper, or a PUT on a person resource with only 1 address should not delete any existing addresses.  Similarly, including only 1 custom field, should not delete all existing custom fields on a resource.
+
+#### Deleting the full contents of an Object (hash) or Array
+In order to cause the deletion of the contents of an Object (hash) or Array, a request should include that element, but set its value to ````null````
```

Comment:
We currently have a known feature gap related to deleting a single item
from an array.  You'd need to get the current addresses, null it out by
setting it to null, then repopulate with the desired addresses.

As far as the PUT case, (or any case really) overwrite semantics are too
dangerous and the result is that no one will ever use the PUT operation.
Can you clarify if you mean:
1) complete overwrite - any missing element, even a top level one like
given_name will result in it being nulled on the server
2) just present subordinate resources overwrite - only an inline object or
array, which is specified will have its data overwritten


On Thu, Jun 11, 2015 at 11:49 AM, Shai Sachs <notifications@github.com>
wrote:

> In README.md
> <https://github.com/opensupporter/osdi-docs/pull/206#discussion_r32235706>
> :
>
> > @@ -306,6 +307,19 @@ Some servers may support certain helpers and not others. The AEP and associated
> >
> >  Similarly, matching behavior will be determined by each implementing system. For example, some systems may match people based on email address or other information.
> >
> > +### Deviations from RESTful Behavior
> > +
> > +This section outlines areas where the expectations of the OSDI customer community differ from boilerplate RESTful behavior and specify best practice expected behavior.
> > +
> > +#### Merging Objects (hashes) and Arrays on update
> > +
> > +Numerous resources in OSDI contain embedded objects (hashes) and arrays. For example, on Person, Birthday and Custom Fields are Objects (hashes) and Postal Addresses is an array.
> > +
> > +When dealing with updates to these elements of resources, servers should merge rather than replace these elements.  This is more consistent with expected user behavior.  For example, sending a person_signup_helper, or a PUT on a person resource with only 1 address should not delete any existing addresses.  Similarly, including only 1 custom field, should not delete all existing custom fields on a resource.
> > +
> > +#### Deleting the full contents of an Object (hash) or Array
> > +In order to cause the deletion of the contents of an Object (hash) or Array, a request should include that element, but set its value to ````null````
>
> I think an example would be useful here, since there's a question in my
> mind about "addressing" the item to be deleted. If a person has three
> addresses and the second is to be deleted, what should the request body
> look like? Kind of along similar lines I'd rather that PUT requests have
> "overwrite" semantics, wrt the text above.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opensupporter/osdi-docs/pull/206/files#r32235706>.
>



-- 

*Josh Co*hen  joshco@gmail.com | @joshco | LinkedIn
<http://linkedin.com/in/joshrcohen/> | GitHub <https://github.com/joshco>
  | OSDI opensupporter.org <http://opensupporter.org/>
"If [OSDI <http://opensupporter.org/>] succeeds, it could usher in the next
chapter in political technology." - Washington Post <http://wapo.st/1hYXX7O>

In the LGBT community, Domestic Abuse is the friendly-fire equivalent of
bullying, sexual assault, discrimination, Religion based bigotry  and other
types of  subjugation.

Let's stop it. Please help
<https://www.crowdrise.com/avpchamps-friends/fundraiser/joshcohen2>